numpy/_core/src/common/npy_argparse.c

```diff
@@ -274,14 +285,22 @@ _npy_parse_arguments(const char *funcname,
         /* ... is NULL, NULL, NULL terminated: name, converter, value */
         ...)
 {
-    if (NPY_UNLIKELY(cache->npositional == -1)) {
-        va_list va;
-        va_start(va, kwnames);
-
-        int res = initialize_keywords(funcname, cache, va);
-        va_end(va);
-        if (res < 0) {
-            return -1;
+    if (NPY_UNLIKELY(!cache->initialized)) {
```

Comment:
This is not thread-safe. You need to use the atomic load below. The atomic load is not going to be slow.